rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+
+- the Vector repo (with the Github Actions based CI flow)
+- the test harness (also integrated with CI, but this is it's own thing)
+
+We don't necessarily have to choose one of those places: we can add a new
+location if it's justified enough.
+
+Let's outline the requirements on the properties of the solution:
+
+- We want to have the ability to run the checks from the Vector repo CI, i.e.
+  per commit, per PR, per tag etc. This might not be immediately utilized, but
+  we just want to have that option.
+
+- We want to consolidate the management of the _cloud_ resources we allocate and
+  pay for Kubernetes test infrastructure in a single place. This is to avoid
+  spreading the responsibility, duplicating the logic, reusing allocated
+  resources for all our testing needs, and simplify accounting and make the
+  configuration management more flexible.
+  We can, for example, have a shared dependency for Vector CI flow, Test Harness
+  invocations, locally run tests - and whatever else we have - to rely on.
+
+- We want our test infrastructure easily available for the trusted developers
+  (Vector core team) to run experiments and tests against locally. This doesn't
+  mean we want to automate this and include running tests locally against our
+  whole k8s test infrastructure - but the ability to do it with little effort is
+  very important: even if we employ super-reliable CI automation, the turnaround
+  time of going through it is way higher than conducting an experiment locally.
+  Locally means using local code tree and binaries - the infrastructure itself
+  is still in the cloud.
+
+- Ideally, we want the test system to be available not just to the Vector core
+  team, but to the whole open-source community. Of course, we don't want to give
+  unrestricted access to _our_ cloud testing infrastructure - but the solution
+  we employ should allow third-parties to bring their own resources. Things that
+  are local in essence (like `minikube`) should just work. There shouldn't be a
+  situation where one can't run tests in `minikube` because cloud parts aren't
+  available. We already have a similar constraints at the Vector Test Harness.
+
+- We need the required efforts to managements the solution to be low, and the
+  price to be relatively small. This means that the solution has to be simple.
+
+- We want to expose the same kind of interface to each of the clusters, so the
+  cluster we run the tests is easily interchangeable.
+  A kubectl config file is a good option, since it encapsulates all the
+  necessary information tp connect to a cluster.
+
+Based on all of the above, it makes sense to split the infrastructure into two
+parts.
+
+- Cloud infrastructure that we manage and pay for.
+
+  We will create a dedicated public repo with [Terraform] configs to setup a
+  long-running Kubernetes test infrastructure.
+  The goal here is to make the real, live cloud environments available for
+  people and automation to work with.
```

Comment:
I'd prefer that we add all of this to the [test harness](https://github.com/timberio/vector-test-harness). What you're describing really isn't much different than https://github.com/timberio/vector-test-harness/pull/44. And this is what the test harness is designed for. Aligning this is important because the more we improve the test harness workflow the more we improve everything associated. For example, the Github action we built makes it easy for us to test Kubernetes integration tests with a Github comment, one that is restricted to Vector team members. I don't want to recreate that workflow for another repo.

Happy to hear reasons to not do this, but I am corned about adding more "parts" to our testing process.